The Failsafe can now recover from an deleted MC

Gamer025 opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
Its also more reliable and can handle a situation where its main Loop runtimes and the MC is stuck
You can also now debug Master/New()

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
While there will most likely never be any situation where the MC is just gone its still good to know that the game can recover from such a situation
For example maybe someone messed up a SDQL query or maybe someone wanted to delete the MC to create a new one hoping the Failsafe would do so for him
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:Gamer025
code: Failsafe is more robust now and the game can actually recover from an deleted MC
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Gamer025 committed:
The Failsafe can now recover from an deleted MC
Its also more reliable and can handle a situation where its main Loop runtimes and the MC is stuck

Gamer025 committed:
Reset defcon level correctly
Oops left that in from debugging the levels

Cyprex commented:
@MrStonedOne

Cyprex commented:
Did you test this? If so, what in particular did you test?

Gamer025 commented:
> Did you test this? If so, what in particular did you test?

I tested all the different stages of the emergency loop multiple times.
All of them but the last one which gets rid of all subsystem were able to pretty much restore the game to working order as far as I could tell.
Like movement worked, atmos was processing, chat was behaving normal and there were no runtimes expect for the last stage where it deletes the SS as a last resort ditch.
In the last case there were a lot of runtimes and some stuff broke but you could still move and the chat was working which I think is a better situation then having no Master which means no movement at all etc.

The game also behaved normal when the MC wasn't deleted because the very first MC pretty much still inits like before.

MrStonedOne commented on code/controllers/failsafe.dm:
```diff
@@ -106,6 +115,41 @@ GLOBAL_REAL(Failsafe, /datum/controller/failsafe)
+					del(global.vars[global_var])
+			. = Recreate_MC()
+
+	if (. == 1) //We were able to create a new master
```
The biggest issue i see with this logic is that recreate_mc will return 0 and not attempt to recreate the mc if it has been 'too soon' since it last attempted to recreate the mc, this should either send an argument to bypass this logic or handle it by not decrementing defcon in those cases.

Gamer025 replied:
Fixed in latest commit
Will now only decrease defcon if Recreate_MC explicitly returns -1

MrStonedOne commented on code/controllers/master.dm:
```diff
@@ -124,7 +141,8 @@ GLOBAL_REAL(Master, /datum/controller/master) = new
 	var/delay = 50 * ++Master.restart_count
```
All of this proc needs to have `Master.` replaced with `Master?.` or it will just runtime

Gamer025 replied:
Isn't restart_count, restart_timeout and restart_clear static, so they will always exist?
That's why I only added the check if Master exists for processing

MrStonedOne replied:
actually im pretty sure this works now because Master is a typed variable and these access static variables. amazing.

LemonInTheDark replied:
what the fuck

Gamer025 replied:
Well in my testing this way of accessing the vars always worked, so I guess its fine the way it currently is

Gamer025 committed:
Correctly recover SSasset

Gamer025 committed:
Only decrease defcon if MC creation failed
Also add some sort sleep between emergency loops

Gamer025 committed:
Makes the last two emergency actions manual procs
Since they are kinda unstantable its probalby best
if only admins call these manually
